rfcs/solidity-snapps.md

```diff
@@ -0,0 +1,324 @@
+current RFC doesn't identify what the proof inputs should be, but describes a
+stack of parties (`parties`) and the current state of the stack when a
+transaction is reached in the stack (`remaining_parties`).
+
+**Proposal:** pass `parties`, the stack of parties, as part of the snapp input.
+
+We should also consider nested function calls, each of which may result in one
+or more parties (e.g. to pay one or more receivers, or to make other further
+function calls). We can make these conceptually simpler and more composable by
+grouping the transactions nested below a party's snapp together, in a hierarchy
+of snapps. This will be particularly helpful for snapps which make recursive
+calls or deeply nested calls, by letting them avoid walking arbitrarily far
+along the stack of parties to find the one they care about.
+
+**Proposal:** use a stack of stacks for the parties involved in a transaction,
+allowing each snapp to access its inner transactions by examining its stack.
+For example, a snapp which calls other snapps might have a stack that looks
+like
+```ocaml
+[ transfer_for_fee
+; [ snapp1
+  ; transfer1 (* Sent by snapp1 *)
+  ; [ snapp2 (* Called by snapp1 *)
+    ; transfer2 (* Sent by snapp2 *)
+    ; [snapp3] ] (* Called by snapp2 *)
+  ; transfer3 (* Sent by snapp1 *)
+  ; [snapp4] ] ] (* Called by snapp1 *)
+```
+Concretely, this allows snapp1 to access `transfer3` and `snapp4` without
+needing to know or care about the transfers and snapps executed by `snapp2`.
+In the implementation, this could look something like:
+```ocaml
+let get_next_party
+    current_stack (* The stack for the most recent snapp *)
+    call_stack (* The partially-completed parent stacks *)
+  =
+  let next_stack, next_call_stack =
+    if call_stack = empty_stack then
+      empty_stack, empty_stack
+    else
+      call_stack.pop()
+  in
+  (* If the current stack is complete, 'return' to the previous
+     partially-completed one.
+  *)
+  let current_stack, call_stack =
+    if current_stack = empty_stack then
+      next_stack, next_call_stack
+    else
+      stack, call_stack
+  in
+  let stack_or_party, next_stack = current_stack.pop() in
+  let party, remaining_stack =
+    let stack =
+      if stack_or_party.is_party() then
+        (* dummy value for circuit *)
+        current_stack
+      else
+        stack_or_party.as_stack()
+    in
+    let popped_value, remaining_stack = stack.pop() in
+    if stack_or_party.is_party() then
+      stack_or_party.as_party(), empty_stack
+    else
+      popped_value, remaining_stack
+  in
+  let party, current_stack, next_stack =
+    if remaining_stack = empty_stack then
+      party, next_stack, empty_stack
+    else
+      party, remaining_stack, next_stack
+  in
+  let call_stack =
+    if next_stack = empty_stack then call_stack
+    else call_stack.push(next_stack)
+  in
+  party, current_stack, call_stack
+```
+
+This increases the number of stack operations per party from 1 to 4.
```

Comment:
This should not trouble us much as a push/pop is just one hash, which is pretty cheap (and will be even cheaper in the new proof system)